src/zarr/core/array.py

```diff
@@ -138,11 +138,43 @@
 
 
 # Array and AsyncArray are defined in the base ``zarr`` namespace
-__all__ = ["create_codec_pipeline", "parse_array_metadata"]
+__all__ = [
+    "DEFAULT_FILL_VALUE",
+    "DefaultFillValue",
+    "create_codec_pipeline",
+    "parse_array_metadata",
+]
 
 logger = getLogger(__name__)
 
 
+class DefaultFillValue:
+    """
+    Sentinel class to indicate that the default fill value should be used.
+
+    This class is used to distinguish between:
+    Zarr Format 3:
+        - fill_value = None: dtype default value
+    Zarr Format 2:
+        - fill_value = None: fill_value saved as null
+
+    This allows backwards compatibility with zarr format 2.
+
+    This is implemented as a singleton.
+    """
+
+    _instance = None
+
+    def __new__(cls) -> Self:
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
```

Comment:
is this perhaps too fancy? I feel like a vanilla class is OK here, without the need for singleton-ness. The code that will use instances of this class is already checking against a specific instance. That seems like all we need.